text/0000-namespace-syscalls.md

```diff
@@ -0,0 +1,75 @@
+- Feature Name: namespace-syscalls
+- Start Date: 2016-11-23
+- RFC PR: (leave this empty)
+- Redox Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+A namespace is designed to implement the following with one abstraction:
+- `cap_enter`, by default
+- `chroot`, by allowing a filter on `file:`
+- [OS-level virtualization](https://en.wikipedia.org/wiki/Operating-system-level_virtualization) such as [FreeBSD-style Jails](https://en.wikipedia.org/wiki/FreeBSD_jail) or [Illumos-style Zones](https://en.wikipedia.org/wiki/Solaris_Containers), with more complex filtering of scheme access
+
+It achieves this with the addition of three syscalls:
+- `getns`, which gets the current namespace
+- `mkns`, which creates a new namespace
+- `setns`, which switches namespaces
+
+# Motivation
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected outcome?
+
+# Detailed design
```

Comment:
A namespace is a kernel construct - an isolated list of schemes, scheme handlers, and scheme consumers.

The scheme is not aware of the caller's namespace, unless it constructed that namespace such as how the `contain` program works.

There is not a hierarchy in the kernel - however, hierarchies can be constructed. The `contain` program will use the `file:` from its original namespace and translate and reexport it in the new namespace.

This affects all contexts, threads and processes, spawned after a `setns`.

The context remembers its original namespace, and can switch between a new namespace that it created and its original namespace. A context spawned after setns cannot switch back to its parent's namespace.